### ISSUE TITLE
HDWalletProvider silently exits when rpc endpoint is not available

### ISSUE THREAD

**Issue by TripleSpeeder** [trufflesuite/truffle#2655]
- [ X ] I've asked for help in the [Truffle Gitter](http://gitter.im/Consensys/truffle) before filing this issue.
-> No luck there :-(

---------------------------

## Issue

Using HDWalletProvider in a node.js application. When the rpc endpoint is not available the whole node process exits silently, instead of throwing an error.

## Steps to Reproduce

Short testcase:
```javascript
onst HDWalletProvider = require('@truffle/hdwallet-provider')
const Web3 = require('web3')

const providerString = 'http://127.0.0.1:8545'

async function checkPlainProvider() {
    console.log("Setting up plain web3")
    const plainWeb3 = new Web3(providerString)
    try {
        console.log("\tTrying to get nodeInfo. Should throw an Error due to invalid provider.")
        let result = await plainWeb3.eth.getNodeInfo()
        console.log("\tResult: " + result)
    } catch (error) {
        console.log("\tError message: " + error.message)
        console.log("\tCaught error as expected. Returning.")
        return
    }
    console.log("\tShould never get here")
}

async function checkHDWalletProvider() {
    console.log("Setting up web3 with HDWalletProvider")
    const hdwalletprovider = new HDWalletProvider('0xaabbcc', providerString)
    const hdWeb3 = new Web3(hdwalletprovider)
    try {
        // This call should throw an exception. But instead the whole process dies without showing any error.
        console.log("\tTrying to get nodeInfo. Should throw an Error due to invalid provider.")
        let result = await hdWeb3.eth.getNodeInfo()
        console.log("\tResult: " + result)
    } catch (error) {
        console.log("\tError message: " + error.message)
        console.log("\tCaught error as expected. Returning.")
        return
    }
    console.log("\tShould never get here...")
}

async function runme() {
    await checkPlainProvider()
    await checkHDWalletProvider()
}

runme()
```

Output:
```
> node providertest.js 
Setting up plain web3
        Trying to get nodeInfo. Should throw an Error due to invalid provider.
        Error message: Invalid JSON RPC response: ""
        Caught error as expected. Returning.
Setting up web3 with HDWalletProvider
        Trying to get nodeInfo. Should throw an Error due to invalid provider.
>
```

## Expected Behavior
An error should be thrown instead of silent exit.

## Actual Results

The process just dies, no exit code set, no exception thrown.

## Environment

* Operating System: Arch Linux
* Truffle v5.1.2 (core: 5.1.2)
* npm 6.13.0
* Web3.js v1.2.2
* Node v10.16.3 (also happens with node 8.x and 12.x)

**Comment by TripleSpeeder** [comment]
Note that this looks similar to #1865 which has been closed due to inactivity.

**Comment by CruzMolina** [comment]
@TripleSpeeder , will need to reproduce & investigate locally, but does the same behavior happen when using Web3.js v1.2.1?

**Comment by CruzMolina** [comment]
Noting here for future investigation: because of possible changes (bundled typings) in Web3.js in 1.2.2 and later versions, we might need to update the `@truffle/hdwallet-provider` docs to state compatible versions of that library with Web3.js.

**Comment by TripleSpeeder** [comment]
> @TripleSpeeder , will need to reproduce & investigate locally, but does the same behavior happen when using Web3.js v1.2.1?

Yes, same behaviour confirmed with web3 1.2.1.

**Comment by CruzMolina** [comment]
Hey @TripleSpeeder , after digging around in the code, it looks like [this](https://github.com/trufflesuite/truffle/blob/develop/packages/hdwallet-provider/src/index.ts#L160-L170) is probably why the `getNodeInfo` call is not failing as expected. IIRC, `web3-provider-engine` (which `@truffle/hdwallet-provider` uses under the hood) mocks a legitimate provider. Not sure what the solution is here off-hand.

**Comment by fabianriewe** [comment]
Hey, I don't know if this is relevant in this issue. I was using the provider as a string which was working for me fine when executing a JS script in node locally. But when moving the code to VueJs, nothing was happening. The HDWallet was not loaded correctly:
```javascript
const endpoint = "http://127.0.0.1:9545"
let hd = new HDWalletProvider(mnemonic, endpoint);
this.web3 = new Web3(hd);
````
But after replacing the string with HttpProvider, everything was working fine:
```javascript
const endpoint = "http://127.0.0.1:9545"
let hd = new HDWalletProvider(mnemonic, new Web3.providers.HttpProvider(endpoint));
this.web3 = new Web3(hd);
````

**Comment by sergiocasero** [comment]
@fabianriewe thank you so much!! You save my day! :)

**Comment by eggplantzzz** [comment]
Hey @TripleSpeeder, thanks for reporting this. It would definitely be better for this to throw an error.
